### ISSUE TITLE
SPEC2017 519.lbm regression on riscv gcc 11 backport of b646d7d279ae 

### ISSUE THREAD

**Issue by vineetgarc** [vineetgarc/riscv-gcc#1]
_Note: This issue involves an upstream gcc patch (for gcc 13, officially backported to gcc-12) but the issue happens on a gcc-11 backport hence no official bugzilla entry or issue on [riscv gcc repo](https://github.com/riscv-collab/riscv-gcc/)_

The gcc change b646d7d279ae ("RISC-V: Inhibit FP <--> int register moves via tune param")to elide FMV instructions can benefit performance on uarches where FMV are costlier and repeated stack spills could potentially be cached. And it is supposed to be codesize neutral - under high register pressure it would replace FMV.x.d/FMV.d.x. pair with FLD/FST pair.

However with a gcc 11 backport of above, we see regression on SPECFP2017 benchmark 519.lbm (dynamic instructions count).
The additional stack spills are obviously there, but there are additional FLD for loading floating point constants from .rodata - multiple times - and a few times the register hoisting the constant is still live.

```
336:		u2 = 1.5 * (ux*ux + uy*uy + uz*uz);

338:		feqs[C ] =            (1.0/3.0)*rho*(1.0                         - u2);
339:         feqs[N ] = feqs[S ] = (1.0/18.0)*rho*(1.0 + 4.5*(+uy )*(+uy ) - u2);

...
```
**codegen BEFORE b646d7d279ae**

```
   10cb8:	8345784b          	fnmsub.d	fa6,fa0,fs4,fa6
   10cbc:	d3b27243          	fmadd.d	ft4,ft4,fs11,fs10
   10cc0:	5b4575cb          	fnmsub.d	fa1,fa0,fs4,fa1
   10cc4:	1236f6d3          	fmul.d	fa3,fa3,ft3
   10cc8:	d3457dcb          	fnmsub.d	fs11,fa0,fs4,fs10
   10ccc:	0b4570cb          	fnmsub.d	ft1,fa0,fs4,ft1
   10cd0:	0345704b          	fnmsub.d	ft0,fa0,fs4,ft0
   10cd4:	2b4572cb          	fnmsub.d	ft5,fa0,fs4,ft5
   10cd8:	3345734b          	fnmsub.d	ft6,fa0,fs4,ft6
   10cdc:	2345724b          	fnmsub.d	ft4,fa0,fs4,ft4
   10ce0:	1345714b          	fnmsub.d	ft2,fa0,fs4,ft2
   10ce4:	9345794b          	fnmsub.d	fs2,fa0,fs4,fs2
   10ce8:	bc42                	fsd	fa6,56(sp)
   10cea:	b42e                	fsd	fa1,40(sp)
   10cec:	ac36                	fsd	fa3,24(sp)
   10cee:	be05                	j	1081e <main+0x2e0>
```

**codegen AFTER b646d7d279ae**

Note the repeated refetch of const @ LC29 = 1.5 used as multiplier in line 336 of src.
```
# lbm.c:338: 		feqs[C ] =            (1.0/3.0)*rho*(1.0                         - u2);
   10c8e:	c316f8cb          	fnmsub.d	fa7,fa3,fa7,fs8
   10c92:	3c06                	fld	fs8,96(sp)
   10c94:	bcc6                	fsd	fa7,120(sp)
   10c96:	8a81b887          	fld	fa7,-1880(gp)              # fld fa7, %lo(.LC29)(a3)
   10c9a:	c316f8cb          	fnmsub.d	fa7,fa3,fa7,fs8
   10c9e:	b8c6                	fsd	fa7,112(sp)
   10ca0:	8a81bc07          	fld	fs8,-1880(gp)              # fld fs8, %lo(.LC29)(a3)
   10ca4:	28e6                	fld	fa7,88(sp)
   10ca6:	f386ff4b          	fnmsub.d	ft10,fa3,fs8,ft10
   10caa:	b4fa                	fsd	ft10,104(sp)
   10cac:	8b86ff4b          	fnmsub.d	ft10,fa3,fs8,fa7
   10cb0:	2c46                	fld	fs8,80(sp)
   10cb2:	8a81b887          	fld	fa7,-1880(gp)              # fld fa7, %lo(.LC29)(a3)
   10cb6:	b0fa                	fsd	ft10,96(sp)
   10cb8:	c316ff4b          	fnmsub.d	ft10,fa3,fa7,fs8
   10cbc:	2886                	fld	fa7,64(sp)          # NOK: could have reused fa7 and fld into fa8
   10cbe:	8a81bc07          	fld	fs8,-1880(gp)              # fld fs8, %lo(.LC29)(a3)
   10cc2:	acfa                    fsd	ft10,88(sp)
   10cc4:	8b86ff4b          	fnmsub.d	ft10,fa3,fs8,fa7
   10cc8:	38c2                	fld	fa7,48(sp)
   10cca:	8a81bc07          	fld	fs8,-1880(gp)              # fld fs8, %lo(.LC29)(a3)
   10cce:	a8fa                    fsd	ft10,80(sp)
   10cd0:	8b86ff4b          	fnmsub.d	ft10,fa3,fs8,fa7
   10cd4:	3c22                	fld	fs8,40(sp)
   10cd6:	a0fa                	fsd	ft10,64(sp)
   10cd8:	8a81bf07          	fld	ft10,-1880(gp)              # fld ft10, %lo(.LC29)(a3)
   10cdc:	c3e6f8cb          	fnmsub.d	fa7,fa3,ft10,fs8
   10ce0:	2c62                	fld	fs8,24(sp)
   10ce2:	b846                	fsd	fa7,48(sp)
   10ce4:	c3e6f8cb          	fnmsub.d	fa7,fa3,ft10,fs8
   10ce8:	2c42                	fld	fs8,16(sp)
   10cea:	a83a                	fsd	fa4,16(sp)
   10cec:	c3e6f6cb          	fnmsub.d	fa3,fa3,ft10,fs8
   10cf0:	b446                	fsd	fa7,40(sp)
   10cf2:	ac36                	fsd	fa3,24(sp)
   10cf4:	b61d                	j	1081a <main+0x2dc>
...
.LC29:
	.word	0
	.word	1073217536	# 1.5e
```
Just before this codegen, there's another such pathetic refetch for a different constant .LC27 1.0 used in lines 339 and beyond.

```
	fld	fa7,%lo(.LC27)(s10)	
	fmadd.d	fa3,fa3,fs4,fa7	
# lbm.c:336: 		u2 = 1.5 * (ux*ux + uy*uy + uz*uz);
	fld	fa7,16(sp)	
	fsd	fs8,104(sp)	
	fmul.d	fs8,fa4,fa4	
	fmul.d	fa4,fa4,fs7	
	fsd	fa3,96(sp)	
	fadd.d	fa3,fa7,fs8
	fld	fa7,%lo(.LC27)(s10)
	fmadd.d	fs8,fs8,fs4,fa7
	fmadd.d	ft10,ft10,fs4,fa7
	fld	fa7,%lo(.LC27)(s10)
	fsd	fs8,88(sp)
	fmul.d	fs8,fs10,fs10	
	fmadd.d	fs8,fs8,fs4,fa7	
	fsd	fs8,80(sp)	
	fmul.d	fs8,fs9,fs9	
	fmadd.d	fa7,fs8,fs4,fa7	
	fld	fs8,24(sp)	
	fsd	fa7,64(sp)	
	fld	fa7,%lo(.LC27)(s10)	
	fmadd.d	fs8,fs8,fs4,fa7	
	fld	fa7,%lo(.LC27)(s10)	
	fsd	fs8,48(sp)	
	fld	fs8,40(sp)	
	fmadd.d	fs8,fs8,fs4,fa7	
	fld	fa7,%lo(.LC27)(s10)	
	fsd	fs8,40(sp)	
	fmul.d	fs8,fs11,fs11	
	fmadd.d	fs8,fs8,fs4,fa7	
	fsd	fs8,24(sp)	
	fld	fs8,104(sp)	
	fmadd.d	fa7,fs8,fs4,fa7	
	fld	fs8,%lo(.LC27)(s10)
	fsd	fa7,16(sp)
...
	.align	3
.LC27:
	.word	0
	.word	1072693248	# 1.0
```

I've yet to create a small test case as this only happens with -flto=auto on final link of benchmark.

**Comment by vineetgarc** [comment]
To debug the issue, I turned fmv_cost from "fixed" cpu_tune into a user toggle:
-fmv-cost=4 (or lesser) exhibits old behavior (FMV.x.d) while -mfmv-cost=6 (or more) which implements to stack spills and exhibits the pathology with additional reloading of const above. This is in the my gcc tree https://github.com/vineetgarc/riscv-gcc/tree/gcc-11-branch

Unfortunately I can't create a smaller reproducer, with enough register pressure to trigger the issue.

**Comment by vineetgarc** [comment]
Started sifting thru the -fdump-rtl-all  and -dP gcc-internals annotated dumps (for -mfmv-cost=6)

```
#(insn 2595 2589 1291 (set (reg:DF 49 fa7 [2364])
#        (mem/u/c:DF (lo_sum:DI (reg:DI 13 a3 [2361])
#                (symbol_ref/u:DI ("*.LC29") [flags 0x82])) [0  S8 A64])) "lbm.c":338:63 146 {*movdf_hardfloat_rv64}
#     (nil))
	fld	fa7,%lo(.LC29)(a3)	# 2595	[c=32 l=8]  *movdf_hardfloat_rv64/2


#(insn 1291 2595 2600 (set (reg:DF 49 fa7 [orig:825 prephitmp_5609 ] [825])
#        (fma:DF (neg:DF (reg:DF 45 fa3 [orig:843 _5653 ] [843]))
#            (reg:DF 49 fa7 [2364])
#            (reg:DF 56 fs8 [2365]))) "lbm.c":338:63 52 {fnmadf4}
#     (expr_list:REG_DEAD (reg:DF 56 fs8 [2365])
#        (nil)))
	fnmsub.d	fa7,fa3,fa7,fs8	# 1291	[c=40 l=4]  fnmadf4
```
The instruction corresponding to constant refetch seems to be introduced in 292r.reload (it is not present in 291r.ira).
It is practically impossible to see whats going.

**Comment by vineetgarc** [comment]
Given that problem oesn't happen on gcc 12.1.0 (but happens on gcc 11) - I decided to git bisect on releases/gcc-12-branch

```
git bisect start --term-new=fixed --term-old=buggy
git bisect fixed 1ea978e3066a    # Update ChangeLog and version files for release (12.1.0)
git bisect buggy 0cc79337ad26  # Bump BASE-VER (12.0.0)
```

I had to manually apply a minimal patch which introduced -mfmv-cost.

After 13 iterations, git bisect pointed to first "fixed" commit

```
4b3a325f07acebf47e82de227ce1d5ba62f5bcae is the first fixed commit
commit 4b3a325f07acebf47e82de227ce1d5ba62f5bcae
Author: Aldy Hernandez <aldyh@redhat.com>
Date:   Thu Oct 28 15:35:21 2021 +0200

    Remove VRP threader passes in exchange for better threading pre-VRP.
    
    This patch upgrades the pre-VRP threading passes to fully resolving
    backward threaders, and removes the post-VRP threading passes altogether.
    With it, we reduce the number of threaders in our pipeline from 9 to 7.
    
    This will leave DOM as the only forward threader client.  When the ranger
    can handle floats, we should be able to upgrade the pre-DOM threaders to
    fully resolving threaders and kill the embedded DOM threader.
    
    The numbers are as follows:
    
            prev: # threads in backward + vrp-threaders = 92624
            now:  # threads in backward threaders = 94275
            Gain: +1.78%
    
            prev: # total threads: 189495
            now:  # total threads: 193714
            Gain: +2.22%
    
            The numbers are not as great as my initial proposal, but I've
            recently pushed all the work that got us to this point ;-).
    
    And... the compilation improves by 1.32%!
```
So updates to Value Range Propagation in gcc 12 release cycle fixed this issue.
